ResolvedTS/ResolvedTS.tla

```diff
@@ -0,0 +1,240 @@
+----------------------------- MODULE ResolvedTS -----------------------------
+EXTENDS Integers, Sequences
+
+\* RM is a set of keys in a region
+CONSTANT RM
+
+VARIABLES
+    rmState, \* State of each region keys.
+    reState, \* ResolvedTS resolver state.
+    nextTS,  \* TSO.
+    resolvedQueue, \* A queue where resolved ts saves.
+    outputTSQueue, \* A queue where commit ts saves.
+    chan,          \* A channel forwards messages to a resolver.
+    chanOffset     \* An offset indicates how many messages have been handled.
+
+vars == <<rmState, reState, nextTS,
+          resolvedQueue, outputTSQueue,
+          chan, chanOffset>>
+
+Messages ==
+    [type: {"Prewrite", "Commit", "Rollback"}, rm: RM, ts: Nat]
+    \cup
+    [type: {"FetchTS"}, ts: Nat]
+
+\* A pair of resolved ts and a index to the outputQueue.
+ResolvedItem == [ts: Nat, index: Nat]
+
+RTTypeOK ==
+    /\ rmState  \in [RM -> [state: {"prewrited", "committed"},
+                           pTS: Nat, cTS: Nat,
+                           msgs: Seq(Messages), appliedOffset: Nat]]
+    /\ reState  \in [RM -> [minPTS: Nat]]
+    /\ nextTS   \in Nat
+    /\ resolvedQueue    \in Seq([ts: Nat, index: Nat])
+    /\ outputTSQueue    \in Seq(Nat)
+    /\ chan             \in Seq(Messages)
+    /\ chanOffset       \in Nat
+
+RTInit ==
+    /\  rmState = [r \in RM |-> [state |-> "committed",
+                   pTS |-> 0, cTS |-> 0,
+                   msgs |-> << >>, appliedOffset |-> 0]]
+    /\  reState = [r \in RM |-> [minPTS |-> 0]]
+    /\  nextTS = 100 \* In practical, nextTS should be a nonzero value.
+    /\  resolvedQueue = << >>
+    /\  outputTSQueue = << >>
+    /\  chan = << >>
+    /\ chanOffset = 0
+
+PessimisticPrewrite(r) ==
+    LET t == nextTS - 2 \* Pessimistic Txns start TS can be stale.
```

Comment:
don't get the point that using `nextTS-2` in this verification,  in a sense, the pessimistic transaction is exactly the same as the optimistic transaction,  for a key TS is always incremental.